crates/rust-analyzer/src/main_loop.rs

```diff
@@ -549,14 +555,35 @@ impl GlobalState {
             }
             // Diagnostics are triggered by the user typing
             // so we run them on a latency sensitive thread.
+            let slice2 = slice.clone();
             self.task_pool.handle.spawn(ThreadIntent::LatencySensitive, {
                 let snapshot = self.snapshot();
                 let subscriptions = subscriptions.clone();
                 move || {
-                    Task::Diagnostics(
+                    Task::Diagnostics(DiagnosticsTaskKind::Syntax(
                         generation,
-                        fetch_native_diagnostics(snapshot, subscriptions, slice),
-                    )
+                        fetch_native_diagnostics(
+                            snapshot,
+                            subscriptions,
+                            slice,
+                            NativeDiagnosticsFetchKind::Syntax,
+                        ),
+                    ))
+                }
+            });
+            self.task_pool.handle.spawn(ThreadIntent::LatencySensitive, {
```

Comment:
This duplicates the amount of tasks we do now, it also spawns semantic and parser diagnostics tasks in lockstep which means we might push queue up some syntax diagnostic tasks after semantic ones. I think a better way to model this might be having the `DiagnosticsTaskKind::Syntax` spawn the corresponding semantic one once its finished computing. That way we make sure to first do the syntax ones.

Reply:
> it also spawns semantic and parser diagnostics tasks in lockstep which means we might push queue up some syntax diagnostic tasks after semantic ones.

That's terrible I couldn't think about it 😅 
I think that with `TaskPool::spawn_with_sender`, I can publish two tasks in order within a single job